Fix the typegetter _prefix tables to use bytes.

jamadden opened the pull request:
Add test cases for this.

Also consistently pass bytes from the test case.

Fixes #6.

I ran into the namespace package issue ("can't import zope.interface") when testing with tox using 'setup.py test' so I switched the runners to zope.testrunner, which seems to have solved the problem.

I took this opportunity to enable coverage reports as well as pip caching, and added Python 3.6 to the support list.

tseaver commented on src/zope/mimetype/tests/__init__.py:
```diff
@@ -0,0 +1,38 @@
+##############################################################################
+"""Test harness for `zope.mimetype` doctests..
+"""
+from __future__ import print_function, absolute_import
```
Not needed.

jamadden replied:
Sometimes, depending on the virtualenv in use and the phase of the moon and what have you, the linter complains, saying something like "zope.foo is a relative import; you should use import zope.mimetype.zope.foo". `absolute_import` fixes that problem.

And when is `print_function` a *bad* idea? I for one don't miss `print >> sys.stderr, "Did I get the syntax right?"` 😄 

So those two are a part of my standard template for files. I don't see what they hurt. If it really  bothers you though I can take them out (and I'd like to understand the harm).

tseaver replied:
Unneeded `__future__` imports are noise:  we deliberately avoid non-absolute imports

1) Your linter is broken if it thinks `zope.foo` is a relative import. (It is also broken if it complains about not assigning the return value of `file.write`).

2) Library / test code shouldn't be printing, ever.  doctests are the obvious exception, but I'd rather shoot them than fix them.

tseaver commented on src/zope/mimetype/tests/test_typegetter.py:
```diff
@@ -0,0 +1,52 @@
+# FOR A PARTICULAR PURPOSE.
+#
+##############################################################################
+from __future__ import print_function, absolute_import
```
Not needed.

tseaver commented on src/zope/mimetype/tests/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+zope.interface.directlyProvides(
+    ISampleContentTypeTwo,
+    zope.mimetype.interfaces.IContentTypeInterface)
```
Interfaces providing interfaces makes my head hurt:  there is something wrong with the design here.

jamadden replied:
It's a pretty engrained part of the way this package works, for good or ill, and would represent a significant change to do otherwise...

tseaver replied:
I wasn't asking for a change here, just "arguing with the ref" for future purposes.

tseaver commented on src/zope/mimetype/tests/test_doctests.py:
```diff
@@ -38,65 +36,45 @@
     return unittest.TestSuite((
         doctest.DocFileSuite(
-                'retrieving_mime_types.txt',
+                '../retrieving_mime_types.txt',
```
Ugh, I hate that this package is tested (except for your new module) solely by doctests.

jamadden replied:
Me too, I'd forgotten that. I just couldn't bring myself to add to them.

Thanks for the review, BTW.

tseaver commented on src/zope/mimetype/typegetter.py:
```diff
@@ -28,7 +28,7 @@ def mimeTypeGetter(name=None, data=None, content_type=None):
     if content_type:
         try:
-            major, minor, params = zope.contenttype.parse.parseOrdered(
+            major, minor, _ = zope.contenttype.parse.parseOrdered(
```
Are you running a linter which complains about the unused variables?

jamadden replied:
Indeed yes. pylint complains by default.

mgedmin replied:
Won't this be fun if zope.mimetype ever starts using `_` for i18n?

I'd rather see `major, minor = ...parseOrdered(...)[:2]`, or a `del params  # shut up pylint warning` or something.

mgedmin commented on setup.py:
```diff
@@ -104,7 +105,7 @@ def read(*rnames):
             'zope.testing',
             ],
-      test_suite='zope.mimetype.tests.test_suite',
+      test_suite='zope.mimetype.tests',
```
If I'm not mistaken, this change means `python setup.py test` is now ignoring doctests.

Since we now have more than one test module and don;t have a single `test_suite()`, we have two options remaining:

- drop `python setup.py test` support
- use `additional_tests()` like in https://github.com/gtimelog/gtimelog/blob/master/src/gtimelog/tests/test_timelog.py#L1558-L1567

jamadden replied:
I didn't know about 'additional_tests'. It seems to work like a charm.

mgedmin commented on src/zope/mimetype/tests/__init__.py:
```diff
@@ -0,0 +1,38 @@
+    """This is a sample content type interface."""
+ISampleContentTypeOne.setTaggedValue("title", u"Type One")
+ISampleContentTypeOne.setTaggedValue("extensions", [])
+ISampleContentTypeOne.setTaggedValue("mimeTypes", ["type/one", "type/foo"])
```
TIL zope interfaces have tagged values.

mgedmin commented on src/zope/mimetype/tests/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+zope.interface.directlyProvides(
+    ISampleContentTypeOne,
+    zope.mimetype.interfaces.IContentTypeInterface)
```
I wonder if you could do this via a class decorator...

mgedmin commented on src/zope/mimetype/tests/test_typegetter.py:
```diff
@@ -0,0 +1,52 @@
+        mimeType = typegetter.mimeTypeGuesser(data=b'<HTML')
+        self.assertEqual('text/html', mimeType)
+
+
```
Too many blank lines here for my taste.

I'm sure [pycodestyle](https://pypi.python.org/pypi/pycodestyle) (nee pep8) would complain.

jamadden replied:
PEP8 actually [says](https://www.python.org/dev/peps/pep-0008/#blank-lines):

> Surround top-level function and class definitions with two blank lines.

That's often ignored/violated though (and is a bit fussy for my taste).

mgedmin replied:
Right, which is why two blank lines in the middle of a method surprised me.

jamadden replied:
Oh. That's not what I thought you were commenting on. My mistake.

mgedmin replied:
My bad -- I didn't consider that GitHub will hide the context following the line I commented on.

mgedmin commented on src/zope/mimetype/typegetter.py:
```diff
@@ -59,9 +59,9 @@ def mimeTypeGuesser(name=None, data=None, content_type=None):
     if data and not mimeType:
         # no idea, really, but let's sniff a few common things:
-        for prefix, type, charset in _prefix_table:
+        for prefix, sniffed_type, _charset in _prefix_table:
```
Wait, is that also for pylint?  So it doesn't complain if the unused local starts with a `_`?  Then `_params` and `_encoding` would be better than my earlier suggestions in the above two pylint fixes.

jamadden replied:
Good point. Done.

mgedmin commented on src/zope/mimetype/typegetter.py:
```diff
@@ -73,12 +73,13 @@ def smartMimeTypeGuesser(name=None, data=None, content_type=None):
 
     if data and mimeType == "text/html":
-        for prefix, type, charset in _xml_prefix_table:
+        for prefix, _, _ in _xml_prefix_table:
```
*sigh*

Maybe namedtuples?

jamadden commented:
I think I've addressed all the feedback provided by both @tseaver and @mgedmin . Thanks very much for the reviews and helpful comments!

Could we please get a merge and release when convenient? This is needed to move forward on zopefoundation/zope.file#1.

tseaver commented:
https://pypi.python.org/pypi/zope.mimetype/2.2.0

jamadden commented:
Excellent, thanks!
